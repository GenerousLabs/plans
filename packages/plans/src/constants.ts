// This is moved out of `store.ts` so as to avoid a circular dependency.
// However, this is something of a hack. It means that we cannot set this value
// while bootstrapping this package. Perhaps there's a more elegant way of
// achieving this...
export const REDUX_ROOT_KEY = '__plans' as const;
