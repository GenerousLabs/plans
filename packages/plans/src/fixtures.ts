import mockFs from 'mock-fs';

export const bob = {
  slug: 'bob',
  data: { name: 'Bob' },
  frontmatter: `---
name: Bob
---`,
  markdown: '',
};
export const charlie = {
  slug: 'charlie',
  data: { name: 'Charlie' },
  frontmatter: `---
name: Charlie
---`,
  markdown: '',
};
export const aliceMessage = {
  data: { sender: 'alice', dateTimestampSeconds: 1590595620 },
  frontmatter: `---
sender: alice
dateTimestampSeconds: 1590595620
---`,
  markdown: 'Are you up for sharing this with me?',
};
export const charlieMessage = {
  data: { sender: 'charlie', dateTimestampSeconds: 1590595860 },
  frontmatter: `---
sender: charlie
dateTimestampSeconds: 1590595860
---`,
  markdown: "Would love to share this with you if you're interested.",
};

export const spotify = {
  slug: 'spotify',
  data: { name: 'Spotify' },
  frontmatter: `---
name: Spotify
---`,
  markdown: `I have two spots on my family plan.
`,
};
export const nordvpn = {
  slug: 'nordvpn',
  data: { name: 'NordVPN' },
  frontmatter: `---
name: NordVPN
---`,
  markdown: `I can use 6 devices at the same time
`,
};
export const omgyes = {
  slug: 'omgyes',
  data: { name: 'OMG Yes' },
  frontmatter: `---
name: OMG Yes
---`,
  markdown: `OMG Yes subscription
`,
};

export const joinFrontmatter = ({
  frontmatter,
  markdown,
}: {
  frontmatter: string;
  markdown: string;
}) => {
  return `${frontmatter}\n${markdown}`;
};

export const mockFilesystem = () => {
  mockFs({
    'alice/': {
      [bob.slug]: {
        'index.md': joinFrontmatter(bob),
        plans: {
          [spotify.slug]: {
            'index.md': joinFrontmatter(spotify),
          },
          [nordvpn.slug]: {
            'index.md': joinFrontmatter(nordvpn),
          },
        },
      },
      charlie: {
        'index.md': joinFrontmatter(charlie),
        plans: {
          [omgyes.slug]: {
            'index.md': joinFrontmatter(omgyes),
            [`message-${aliceMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              aliceMessage
            ),
            [`message-${charlieMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              charlieMessage
            ),
          },
        },
      },
    },
  });
};

export const mockFilesystemRestore = () => {
  mockFs.restore();
};
