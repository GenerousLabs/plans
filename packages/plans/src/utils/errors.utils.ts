export type SerializableError = {
  code?: string;
  data?: any;
  message: string;
};

export const getSerializableError = (error: Error): SerializableError => {
  const { message } = error;
  const code =
    typeof (error as any).code === 'string' ? (error as any).code : undefined;
  const data =
    typeof (error as any).data !== 'undefined'
      ? (error as any).data
      : undefined;
  return { message, code, data };
};
