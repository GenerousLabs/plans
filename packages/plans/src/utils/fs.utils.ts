import { FS } from '../shared.types';

export const doesDirectoryExist = async ({
  fs,
  path,
}: {
  fs: FS;
  path: string;
}) => {
  const stat = await fs.promises.stat(path);
  return stat.isDirectory();
};

export const doesFileExist = async ({ fs, path }: { fs: FS; path: string }) => {
  const stat = await fs.promises.stat(path);
  return stat.isFile();
};
