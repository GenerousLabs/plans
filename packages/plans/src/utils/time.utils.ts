export const timestampSeconds = () => {
  return Math.round(Date.now() / 1e3);
};
