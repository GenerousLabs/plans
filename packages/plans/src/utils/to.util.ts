export const to = <T>(
  promise: Promise<T>
): Promise<
  { result: T; error: undefined } | { result: undefined; error: Error }
> =>
  promise.then(
    (result: T) => ({
      result,
      error: undefined,
    }),
    (error: Error) => ({
      error,
      result: undefined,
    })
  );
