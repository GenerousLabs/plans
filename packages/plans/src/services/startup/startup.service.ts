import isomorphicGitHttp from 'isomorphic-git/http/node';
import { fs } from 'memfs';
import { ME_MOCK_REMOTE } from '../../shared.constants';
import { FS, GitParams } from '../../shared.types';
import { createStore } from '../../store';
import { startup } from './actions/startup.action';

export type RootConfig = {
  // A folder on disk to serve as a root for our set of repositories
  path: string;
  // The remote URL to fetch the "me" repo from
  meRepoRemote: string;
  // This allows us to authenticate against repo remotes
  meRepoHeaders?: {
    [x: string]: string;
  };
};

const getMemfs = () => {
  return (fs as unknown) as FS;
};

export const startWithLocalStore = async ({
  fs,
  http,
  rootConfig,
}: Omit<GitParams, 'dir' | 'headers'> & {
  rootConfig: RootConfig;
}) => {
  const {} = rootConfig;

  const store = createStore();
  await store.dispatch(
    startup({
      fs,
      http,
      rootConfig,
    })
  );

  return store;
};

if (process.env.NODE_ENV === 'development') {
  if (typeof process.env.BOOT_LOCAL_STORE === 'string') {
    console.log('Starting with local store / memfs #2Yc8PN');

    const run = async () => {
      const http = isomorphicGitHttp;
      const fs = getMemfs();
      await fs.promises.mkdir('/connections/');

      // const store = await startWithLocalStore({
      await startWithLocalStore({
        fs,
        http,
        rootConfig: {
          path: '/',
          meRepoRemote: ME_MOCK_REMOTE,
        },
      });
    };

    run();
  }
}
