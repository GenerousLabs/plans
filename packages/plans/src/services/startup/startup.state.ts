import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../store';
import { REDUX_ROOT_KEY } from '../../constants';

export const REDUCER_KEY = 'startup' as const;

type StartupState = {
  initStarted: boolean;
  initCompleted: boolean;
  myUsername?: string;
};

const initialState: StartupState = {
  initStarted: false,
  initCompleted: false,
  myUsername: 'alice',
};

const startupSlice = createSlice({
  name: 'PLANS/startup',
  initialState,
  reducers: {
    initStarted: state => {
      state.initStarted = true;
    },
    initFinished: state => {
      state.initCompleted = true;
    },
    setUsername: (state, action: PayloadAction<string>) => {
      state.myUsername = action.payload;
    },
  },
});

export const { initFinished, initStarted } = startupSlice.actions;

export default startupSlice.reducer;

export const getMyUsername = (state: RootState) =>
  state[REDUX_ROOT_KEY][REDUCER_KEY].myUsername;
export const getMyUsernameOrThrow = (state: RootState) => {
  const myUsername = getMyUsername(state);
  if (typeof myUsername !== 'string') {
    throw new Error('Failed to get my username. #4Qv0OI');
  }
  return myUsername;
};
