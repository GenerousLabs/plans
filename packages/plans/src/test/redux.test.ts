import { describe, expect, it } from 'jest-without-globals';
import * as f from '../fixtures';
import { startup } from '../services/startup/actions/startup.action';
import { ME_MOCK_REMOTE } from '../shared.constants';
import '../utils/time.utils';

const TIMEOUT = 60e3;

let start = 1591281210;
jest.mock('../utils/time.utils', () => {
  return {
    timestampSeconds: () => {
      const d = new Date(start);
      // Add 10s to each call
      start = start + 10;
      return Math.round(d.getTime() / 1e3);
    },
  };
});

describe('redux e2e', () => {
  describe('init', () => {
    it(
      'boots without error #4iRln0',
      async () => {
        const { fs, http, store } = await f.mocksWithClonedMeRepo();

        // NOTE: Jest automatically fails the test if it throws
        await store.dispatch(
          startup({
            fs,
            http,
            rootConfig: {
              meRepoRemote: ME_MOCK_REMOTE,
              path: '/e2e',
            },
          })
        );
      },
      TIMEOUT
    );

    it(
      'Produces a consistent state #uvVYtC',
      async () => {
        const { fs, http, store } = await f.mocksWithClonedMeRepo();

        await store.dispatch(
          startup({
            fs,
            http,
            rootConfig: {
              meRepoRemote: ME_MOCK_REMOTE,
              path: '/e2e',
            },
          })
        );

        expect(store.getState()).toMatchSnapshot();
      },
      TIMEOUT
    );
  });
});
