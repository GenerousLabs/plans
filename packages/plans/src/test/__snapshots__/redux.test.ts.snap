// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux e2e init Produces a consistent state #uvVYtC 1`] = `
Object {
  "__plans": Object {
    "me": Object {
      "myPlans": Object {
        "entities": Object {
          "/e2e/myplans/plans/alice/example-plan/index.md": Object {
            "description": "again
",
            "id": "/e2e/myplans/plans/alice/example-plan/index.md",
            "name": "Example Plan",
            "path": "/e2e/myplans/plans/alice/example-plan",
            "slug": "example-plan",
            "userId": "me",
          },
          "/e2e/myplans/plans/alice/prime/index.md": Object {
            "description": "Amazon Prime. Germany only.
",
            "id": "/e2e/myplans/plans/alice/prime/index.md",
            "name": "Prime",
            "path": "/e2e/myplans/plans/alice/prime",
            "slug": "prime",
            "userId": "me",
          },
        },
        "ids": Array [
          "/e2e/myplans/plans/alice/example-plan/index.md",
          "/e2e/myplans/plans/alice/prime/index.md",
        ],
      },
      "repos": Object {
        "entities": Object {
          "bob": Object {
            "currentHeadCommitHash": "73d2e91ee2a0d24452b5811e401fe79896f02361",
            "id": "bob",
            "lastFetchTimestampSeconds": 1591281,
            "name": "Bob",
            "path": "/e2e/repos/bob",
            "remote": "http://localhost:8174/bob.git",
            "slug": "bob",
          },
          "charlie": Object {
            "currentHeadCommitHash": "32c9584d3e33993fa8af989bef8b8101312935cd",
            "id": "charlie",
            "lastFetchTimestampSeconds": 1591281,
            "name": "Charlie",
            "path": "/e2e/repos/charlie",
            "remote": "http://localhost:8174/charlie.git",
            "slug": "charlie",
          },
        },
        "ids": Array [
          "bob",
          "charlie",
        ],
      },
    },
    "plans": Object {
      "plans": Object {
        "entities": Object {
          "/e2e/repos/bob/plans/bob/nordvpn/index.md": Object {
            "description": "My NordVPN account doesn't get so much use, happy to share.
",
            "id": "/e2e/repos/bob/plans/bob/nordvpn/index.md",
            "name": "NordVPN",
            "path": "/e2e/repos/bob/plans/bob/nordvpn",
            "slug": "nordvpn",
            "userId": "/e2e/repos/bob/plans/bob",
          },
          "/e2e/repos/bob/plans/bob/spotify/index.md": Object {
            "description": "I have a spotify account in DE and can share it.

",
            "id": "/e2e/repos/bob/plans/bob/spotify/index.md",
            "name": "Spotify",
            "path": "/e2e/repos/bob/plans/bob/spotify",
            "slug": "spotify",
            "userId": "/e2e/repos/bob/plans/bob",
          },
          "/e2e/repos/charlie/plans/charlie/pia/index.md": Object {
            "description": "I have a VPN that allows multiple devices, share it...
",
            "id": "/e2e/repos/charlie/plans/charlie/pia/index.md",
            "name": "Private Internet Access",
            "path": "/e2e/repos/charlie/plans/charlie/pia",
            "slug": "pia",
            "userId": "/e2e/repos/charlie/plans/charlie",
          },
        },
        "ids": Array [
          "/e2e/repos/bob/plans/bob/nordvpn/index.md",
          "/e2e/repos/bob/plans/bob/spotify/index.md",
          "/e2e/repos/charlie/plans/charlie/pia/index.md",
        ],
      },
      "users": Object {
        "entities": Object {
          "/e2e/repos/bob/plans/bob": Object {
            "description": "Will be loaded soon",
            "id": "/e2e/repos/bob/plans/bob",
            "name": "Name Coming Soon",
            "path": "/e2e/repos/bob/plans/bob",
            "repoId": "bob",
            "slug": "bob",
          },
          "/e2e/repos/charlie/plans/charlie": Object {
            "description": "Will be loaded soon",
            "id": "/e2e/repos/charlie/plans/charlie",
            "name": "Name Coming Soon",
            "path": "/e2e/repos/charlie/plans/charlie",
            "repoId": "charlie",
            "slug": "charlie",
          },
        },
        "ids": Array [
          "/e2e/repos/bob/plans/bob",
          "/e2e/repos/charlie/plans/charlie",
        ],
      },
    },
    "startup": Object {
      "initCompleted": true,
      "initStarted": true,
      "myUsername": "alice",
    },
  },
}
`;
